### PR TITLE
feat!: convert into pipeline extension

### LINKED ISSUE #131: Provide this library as an Antora 3 pipeline extension
Currently, this library is available as a custom site generator (through https://github.com/Mogztter/antora-site-generator-lunr).

With the introduction of the pipeline extension feature in Antora 3, we can simplify the integration.
The pipeline extension provides a lightweight event-based system that we can leverage to generate a search index.

I think we should register our extension on `beforePublish`:

```js
module.exports.register = (pipeline, { config }) => {
  pipeline
    .on('beforePublish', ({ playbook, contentCatalog }) => {
      // generate a search index
    })
}
```

Additionally, extensions can be configured: https://docs.antora.org/antora/3.0/extend/pipeline/configure-extension/ so we could potentially address the following issues:

- https://github.com/Mogztter/antora-lunr/issues/33
- https://github.com/Mogztter/antora-lunr/issues/75

### PR DESCRIPTION
**Pull request by thor**
Fixes #131.

Exports a register function allowing for this to register as a pipeline
extension with Antora.

Alters the signature of generateIndex as it no longer is required
together with antora-site-generator-lunr.

Requires substantially more work, although mostly in terms of:

- [x] Validating if this is a good approach
- [x] Validating how to handle options (what about environment vars?)
- [x] Updating the tests (using mocked `ContentCatalog`)
 (I've started tweaking the tests locally, but I haven't done much. If it seems applicable I thought I'd at the very least merge in the pages into the content catalogue, and alternatively see if importing the `ContentCatalog` from Antora could be an option if not too much scaffolding.)
- [x] Remove `generateIndex` as exported function

### CODE REVIEW

File: lib/generate-index.js [hunk 1]
```diff
@@ -78,6 +76,9 @@ function generateIndex (playbook, pages, contentCatalog, env) {
         $(this).remove()
       })
 
+      // don't index the table of contents, as this is redundant
+      $('#toc', article).remove()
+
```
Comment (thor) [inline lib/generate-index.js:81]: This was a distraction. I'll either put this in another commit or even better: an entirely separate PR.

**Comment by Mogztter** [comment]
Thanks!

It already looks good!
However, I don't know if I will be able to do a thorough review since I haven't used the pipeline extension API yet.

I will try to take some time next week to play a bit more with the API.

@mojavelinux when you have time, feel free to jump into the discussion/review! 😄 

> Validating how to handle options (what about environment vars?)

What about the built-in mechanism to configure an extension (as described in: https://docs.antora.org/antora/3.0/extend/pipeline/configure-extension/)?

> Updating the tests (consider replacing mock content catalog with real ContentCatalog)

Unless Dan strongly advises us not to do it, I would like to keep using unit tests with mocks as it allows us to test edge cases. 
Having said that, we can also add integration tests (I already done that in: https://github.com/Mogztter/antora-site-generator-lunr/tree/main/test).

> Fixing the tests

👍🏻 

> Consider rewriting the signatures of generateIndex, perhaps dividing or combining things to a larger extent?

Yes we can break compatibility, this will be a major version.
Speaking of which, I think we can remove `module.exports = require('./generate-index')`.

File: lib/extension.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+) {
+  const logger = pipeline.require('@antora/logger').get(packageName)
+
+  if (Object.keys(extraArguments).length > 0) {
+    const args = Object.keys(extraArguments)
+      .map((x) => `"${x}"`)
+      .join(', ')
+    logger.fatal(`Remove unrecognised extension option(s): ${args}`)
+    pipeline.halt()
+  }
```
Comment (Mogztter) [inline lib/extension.js:19]: I'm not sure about that, could you please explain what is the purpose?

File: lib/generate-index.js (hunk 1 above)
Comment (Mogztter) [reply lib/generate-index.js:81]: > or even better: an entirely separate PR.

Yes that's a good idea

File: lib/extension.js (hunk 2 above)
Comment (thor) [reply lib/extension.js:19]: @Mogztter The intention is to prevent spelling mistakes or entirely invalid parameters to the pipeline extension from being silently ignored. It's mostly the former scenario I want to prevent, such as:
```yaml
pipeline:
  extensions:
    - require: antora-lunr
      langugaes: ['no', 'en']
```
silently passing through, with the user not understanding why the index hasn't changed. Additionally, it prevents old or deprecated configuration parameters from remaining in use after they're no longer valid.

WDYT, @Mogztter? Would you prefer a warning?

---

Per the behaviour of `pipeline.halt` and Dan's suggestions, I'd change it to this.
```suggestion
  if (Object.keys(extraArguments).length > 0) {
    const args = Object.keys(extraArguments)
      .map((x) => `"${x}"`)
      .join(', ')
    throw new Error(`Unrecognised extension option(s) for ${extensionName}: ${args}`)
  }
```

**Comment by thor** [comment]
> It already looks good!
> However, I don't know if I will be able to do a thorough review since I haven't used the pipeline extension API yet.
> 
> I will try to take some time next week to play a bit more with the API.

No worries! The API and its use is confined to `extension.js` as it currently is, with the entrypoint being the `register` function. 👍 

> > Validating how to handle options (what about environment vars?)
> 
> What about the built-in mechanism to configure an extension (as described in: https://docs.antora.org/antora/3.0/extend/pipeline/configure-extension/)?

That's the one I'm using! :) Currently, I've more or less removed the environment variables in their entirety and replaced them with the `config`. If that sounds sufficient, then they'll stay removed.

> > Updating the tests (consider replacing mock content catalog with real ContentCatalog)
> 
> Unless Dan strongly advises us not to do it, I would like to keep using unit tests with mocks as it allows us to test edge cases.
> Having said that, we can also add integration tests (I already done that in: https://github.com/Mogztter/antora-site-generator-lunr/tree/main/test).

Alright! If so, I think the addition of `getPages()` to `mockContentCatalog` should suffice.

> > Consider rewriting the signatures of generateIndex, perhaps dividing or combining things to a larger extent?
> 
> Yes we can break compatibility, this will be a major version.
> Speaking of which, I think we can remove `module.exports = require('./generate-index')`.

Sounds good! :+1: I'll do that and import it (and `createIndexFile`) from `extension.js`.

File: lib/extension.js (hunk 2 above)
Comment (mojavelinux) [reply lib/extension.js:19]: I like the idea of validating the config options and of throwing an error if an invalid option is used. We want to make sure the user is on the right path.

File: lib/generate-index.js [hunk 3]
```diff
@@ -12,22 +12,22 @@ const { decode } = require('html-entities')
...
-function generateIndex (playbook, pages, contentCatalog, env) {
+function generateIndex (playbook, contentCatalog, { languages, indexLatestOnly }) {
   let siteUrl = playbook.site.url
   if (!siteUrl) {
     // Uses relative links when site URL is not set
     siteUrl = ''
   }
   if (siteUrl.charAt(siteUrl.length - 1) === '/') siteUrl = siteUrl.substr(0, siteUrl.length - 1)
-  if (!pages.length) return {}
+  if (!contentCatalog.getPages().length) return {}
```
Comment (mojavelinux) [inline lib/generate-index.js:27]: It's possible for this method to return pages which are not publishable. This list needs to be filtered as follows:

```js
contentCatalog.getPages((page) => page.out)
```

I also think this method should only be called once. So grab the documents first, then check if the collection is empty.

**Review by mojavelinux — CHANGES_REQUESTED** [review]

**Comment by mojavelinux** [comment]
> Unless Dan strongly advises us not to do it, I would like to keep using unit tests with mocks as it allows us to test edge cases.

That's really up to you. What I can tell you is that we started out using a mock catalog for tests in Antora itself, and I regret it. But that's because Antora needs to test its own functionality thoroughly. For extension tests, you only care about interface, so the mock is more suitable. But there is always that slight unknown that maybe it works with the mock, but won't work with the real object. Use your best judgement and experience will guide you ;)

**Comment by mojavelinux** [comment]
> Currently, I've more or less removed the environment variables in their entirety and replaced them with the config. If that sounds sufficient, then they'll stay removed.

I love this. Progress!

**Comment by mojavelinux** [comment]
I think this looks great. Definitely a huge step forward. Thanks Thor!

**Comment by mojavelinux** [comment]
As I said in the issue, we may want to merge this into a dedicated branch that will become the main branch when the project graduates.

File: lib/generate-index.js (hunk 3 above)
Comment (thor) [reply lib/generate-index.js]: Thanks for both of those, @mojavelinux. Done!

File: lib/extension.js (hunk 2 above)
Comment (thor) [reply lib/extension.js]: I've changed it to throw an error rather than halt the pipeline quietly.

File: lib/generate-index.js (hunk 1 above)
Comment (thor) [reply lib/generate-index.js]: Removed from this PR. :+1:

**Comment by thor** [comment]
I've updated the pull request with the requested changes, and now I only hope I didn't miss something. :crossed_fingers: The tests run well locally, but it'd be great to get the workflow approved in order to see that the tests indeed also work in the pipeline. I have yet to do a clean project build to attempt to verify that.

WDYT @Mogztter? :)

File: .eslintrc.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (mojavelinux) [inline .eslintrc.json:1]: I would prefer if you aligned with the eslint configuration used in Antora core. See https://gitlab.com/antora/antora/-/blob/master/.eslintrc

File: lib/extension.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+function register (pipeline, {
+  config: { indexLatestOnly, languages, ...extraArguments }
+}) {
+  const logger = pipeline.require('@antora/logger').get(packageName)
+
+  if (Object.keys(extraArguments).length > 0) {
+    const args = Object.keys(extraArguments)
+      .map((x) => `"${x}"`)
+      .join(', ')
+    throw new Error(`Remove unrecognised extension option(s) for ${packageName}: ${args}`)
```
Comment (mojavelinux) [inline lib/extension.js:16]: Please use American English spelling in error messages, which means change "unrecognised" to "unrecognized".

Comment (thor) [reply lib/extension.js:16]: Oof, that stings. However, fair enough.

**Comment by Mogztter** [comment]
What I plan to do:

1. Create a new branch named `pipeline` in this repository
1. Update this pull request to target this new branch
1. Review one last time (everything looks good but just to make sure that we didn't mess something during this migration)
1. Merge the pull request
1. Push the repository to https://gitlab.com/antora/antora-lunr-extension
   * `main` branch will become `legacy` in the new repository
   * `pipeline` branch will become `main` in the new repository

@mojavelinux  @thor what do you think? I won't be able to do it this weekend (as I have other plans) but if there are no objections I will do it early next week.

**Comment by thor** [comment]
@Mogztter That sounds good to me. Let me know if you'd like a pull request (or merge request in this case) with `semantic-release`. It's pretty straightforward, so it'd only be to help save you some time.

Speaking of releases, would that perhaps make a `1.0.0` release for `@antora/lunr-extension` once moved?

**Comment by mojavelinux** [comment]
That's a good plan. Not that it matters much, but the name "extension" for the branch seems more appropriate.

> Speaking of releases, would that perhaps make a 1.0.0 release for @antora/lunr-extension once moved?

I agree that starting with 1.0.0 is the right way to go. Starting with pre-1.0.0 releases just puts you in a weird spot with regards to semantic versioning. There are an infinite number of major numbers and we should make use of them ;)

**Comment by Mogztter** [comment]
> That sounds good to me. Let me know if you'd like a pull request (or merge request in this case) with semantic-release. It's pretty straightforward, so it'd only be to help save you some time.

We are not using `semantic-release` in the Asciidoctor/Antora community but since this project is relatively small I don't mind giving it a try. Besides, we will need to migrate to GitLab CI so we could take this opportunity to revisit the build/release process by using `semantic-release`.

@mojavelinux do you think this is a good idea? or do you think that, since the project will be in the `@antora` organization, it should use the same development/build/release workflow (or at least a similar one) as the other projects in the `@antora` organization?

> Speaking of releases, would that perhaps make a 1.0.0 release for @antora/lunr-extension once moved?

Yes that's a good idea 👍🏻 

> That's a good plan. Not that it matters much, but the name "extension" for the branch seems more appropriate.

Noted 👍🏻 

> I agree that starting with 1.0.0 is the right way to go. Starting with pre-1.0.0 releases just puts you in a weird spot with regards to semantic versioning.

👌🏻

**Comment by mojavelinux** [comment]
> @mojavelinux do you think this is a good idea?

I'm not opposed to trying new things. Personally, release libraries just don't work for me. I find myself spending more time trying to make them do what I want them to do that it would take for me to just do it. That's how I arrived at the release process we have now for Antora. But if a library can take on that work, and get it right, then I can't argue with that.

File: package-lock.json [hunk 6]
```diff
@@ -1,609 +1,1766 @@
 {
   "name": "antora-lunr",
   "version": "0.8.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
Comment (Mogztter) [inline package-lock.json:4]: Could you please install the dependencies using npm@6 untill we migrate to npm@7?

File: package.json [hunk 7]
```diff
@@ -33,8 +33,14 @@
...
   },
   "devDependencies": {
+    "@antora/site-generator-default": "^3.0.0-alpha.9",
     "chai": "^4.1.2",
     "dirty-chai": "^2.0.1",
+    "eslint": "^7.32.0",
+    "eslint-config-standard": "^16.0.3",
+    "eslint-plugin-import": "^2.24.2",
+    "eslint-plugin-node": "^11.1.0",
+    "eslint-plugin-promise": "^4.2.1",
```
Comment (Mogztter) [inline package.json:43]: I think we should not add a linter as part of this pull request. I'm in favor of adding a linter (eslint or standard) but once the migration is done.

Comment (mojavelinux) [reply package.json:43]: Agreed. And when we do add it, we should align with the linter configuration used in Antora core (which is open for refinement, of course).

Comment (thor) [reply package.json:43]: Sounds good, @Mogztter. I'll remove it from the PR. :+1:

@mojavelinux I believe the latest version is more (or less?) aligned after the previous update to the PR.

Comment (mojavelinux) [reply package.json:43]: Indeed. I was just mentioning it to clearly state the goal.

**Comment by thor** [comment]
@Mogztter Downgraded to NPM 6 after removing the ESLint packages, as well as its configuration.

**Review by Mogztter — APPROVED** [review]
Awesome work, thanks @thor 🎉 

Sorry for the delay, next step is to push the `extension` branch to https://gitlab.com/antora/antora-lunr-extension and recreate the CI using GitLab CI.

If you are willing to help with that, feel free to open a merge request on GitLab to remove the `.github` and create the `.gitlab-ci.yml`.
